[Training Manual] 3/3 Chapter 8 (Raster) (#3202)
qgis/QGIS-Documentation: source/docs/training_manual/rasters/terrain_analysis.rst

|basic| |FA| Calculating a Hillshade
-------------------------------------------------------------------------------

We are going to use the same DEM layer of the previous lesson, if you are starting
this chapter from scratch use the :guilabel:`Browser` panel and load the
:file:`raster/SRTM/srtm_41_19.tif`.

The DEM layer show you the elevation of the terrain, but it can sometimes seem a
little abstract. It contains all the 3D information about the terrain that you
need, but it doesn't look like a 3D object. To get a better look at the terrain,
it is possible to calculate a *hillshade*, which is a raster that maps the
terrain using light and shadow to create a 3D-looking image.

We are going to use algorithms of :menuselection:`Processing --> Raster terrain analysis`
menu. There are other tools available (other Processing algorithms, plugins, etc.).

* Click on the :menuselection:`Hillshade` menu;

> * The algorithm allows basically where the light source comes from: the

Reviewer: ```suggestion
* The algorithm allows you to specify where the position of the light source: the
```

look better. The key to this is setting the hillshade to being mostly
transparent.

* Change the symbology of the original :guilabel:`srtm_41_19` layer to use the
  :guilabel:`Pseudocolor` scheme as in the previous exercise;
* Hide all the layers except the :guilabel:`srtm_41_19` and :guilabel:`hillshade`
  layers;
* Click and drag the :guilabel:`srtm_41_19` to be beneath the :guilabel:`hillshade`
  layer in the :guilabel:`Layer Styling` panel;
* Set the :guilabel:`hillshade` layer to be transparent by clicking on the

>   :guilabel:`Transparency` tab;

Reviewer: ```suggestion
  :guilabel:`Transparency` tab in the layer properties;
```

example, you want to build houses on the land there, then you need land
that is relatively flat.

To do this, you need to use the :menuselection:`Slope` algorithm of the
:menuselection:`Processing --> Raster terrain analysis`.

* Open the algorithm;
* Choose :guilabel:`srtm_41_19` as the :guilabel:`Elevation layer`;
* Save the output as a file with the name ``slope`` in the same folder of the
  ``hillshade``;
* Click on :guilabel:`Run`.

Now you'll see the slope of the terrain, with black pixels being flat terrain
and white pixels, steep terrain:

.. image:: img/slope_raster.png
   :align: center

.. _backlink-raster-analysis-1:

|moderate| |TY| Calculating the aspect
-------------------------------------------------------------------------------

> The *aspect* of terrain refers to the direction it's facing in. In other words,

Reviewer: ```suggestion
*Aspect* is the compass direction that a slope faces.
```

> you can understand areas facing North, East, South or West.

Reviewer: An aspect of 0 means that the slope is north-facing, 90 east-facing, 180 south-facing, and 270 west-facing.

> you can understand areas facing North, East, South or West.

Reviewer: ```suggestion
An aspect of 0 means that the slope is north-facing, 90 east-facing, 180
south-facing, and 270 west-facing.```

The answer lies with the :guilabel:`Raster calculator`.

QGIS has different raster calculators available:

#. :menuselection:`Raster --> Raster Calculator`;
#. :menuselection:`Processing --> Raster Analysis --> Raster calculator`;
#. :menuselection:`Processing --> GDAL --> Raster miscellaneous --> Raster calculator`;
#. :menuselection:`SAGA --> Raster calculus --> Raster calculator`.

> Each tool is leading to the same results, but the syntax could be slightly

Reviewer: ```suggestion
Each tool is leading to the same results, but the syntax may be slightly
```

> different or some operators could not be available for everyone.

Reviewer: ```suggestion
different and the availability of operators may vary.
```

Reviewer 2: Using Processing>Raster analysis>Raster calculator I seem to get invalid output 
(from information: STATISTICS_VALID_PERCENT=0). Anyone else have that issue? Bug? I did manage to get it to work using Raster>Raster calculator

Reviewer: No problems here with a similar expression as that given in the text ("aspect@1" <= 90 OR "aspect@1" >= 270).
It that the formulae that you have been testing?

Reviewer 2: Yes - I have been investigating a bit: using the raster calculator from processing, I cannot get any results using the aspect layer as input (not even "aspect@1" > 90) - it processes it, creates an output file, which then shows at information:

Name
Output
Path
C:\Users\LENOVO\AppData\Local\Temp\processing_f80613e6691f4551853d505699f4bb71\46c27bdba1414bdcbac0126f66686def\OUTPUT.tif
CRS
EPSG:32733 - WGS 84 / UTM zone 33S - Projected
Extent
969491.2754036566475406,6196099.3408489627763629 : 1038119.7731296116253361,6250296.9955644421279430
Unit
meters
Width
837
Height
661
Data type
Float32 - Thirty two bit floating point 
GDAL Driver Description
GTiff
GDAL Driver Metadata
GeoTIFF
Dataset Description
C:/Users/LENOVO/AppData/Local/Temp/processing_f80613e6691f4551853d505699f4bb71/46c27bdba1414bdcbac0126f66686def/OUTPUT.tif
Compression

Band 1
STATISTICS_VALID_PERCENT=0
More information
AREA_OR_POINT=Area
Dimensions
X: 837 Y: 661 Bands: 1
Origin
969491,6.2503e+06
Pixel Size
81.9934,-81.9934

When I copy and paste the same expression in the raster calculator under menu Raster, it works as expected (values of 0 and 1). With another raster I can use the processing raster calculator, so it seems to be the aspect raster as calculated using the processing toolbox - aspect. Did you use the aspect raster?

Reviewer: I did not use the aspect raster.

We will use :menuselection:`Processing --> Raster Analysis --> Raster calculator`.

* Open the tool by double clicking on it;
* The upper left part of the dialog lists all the raster layers loaded in the
  legend as ``name@N`` where ``name`` is the name of the layer and ``N`` is the
  raster band used;
* In the upper right part you will see a lot of different operators: stop for a
  moment to think that a raster is an image, you should see it as a 2D matrix
  filled with numbers;
* North is at 0 (zero) degrees, so for the terrain to face north, its aspect
  needs to be greater than 270 degrees and less than 90 degrees. Therefore the
  formula is:

  ``aspect@1 <= 90 OR aspect@1 >= 270``

* You have now to set up the raster details, like the cell size, extent and CRS.

>   This can be done manually by filling each different parameter or it can be

Reviewer: ```suggestion
  This can be done manually or it can be
```

  automatically set by choosing a ``Reference layer``. Choose this last option
  by clicking on the |browseButton| of the :guilabel:`Reference layer(s)` parameter;

> * In the dialog choose the :guilabel:`aspect` layer because we want to obtain

Reviewer: ```suggestion
* In the dialog, choose the :guilabel:`aspect` layer because we want to obtain
```

  a layer with the same resolution;
* Save the layer as ``aspect_north``;
* The dialog should look like:

  .. image:: img/raster_calculator.png
     :align: center

* Finally click on :guilabel:`Run`.

Your result will be this:

.. image:: img/aspect_result.png
   :align: center

The output values are ``0`` or ``1``. What does it mean? The formula we wrote

> contains the *conditional* operator ``OR``: therefore the final result will be

Reviewer: ```suggestion
contains *conditionals* and the *conditional* operator ``OR``: therefore the final result will be
```

This operation could be achieved with the raster calculator but the formula would
become very very large.

The alternative tool is the :guilabel:`Reclassify by table` tool within
:menuselection:`Processing --> Raster analysis`.

* Open the tool;
* Choose :guilabel:`aspect` as the ``Input raster layer``;
* Click on the |browseButton| of the :guilabel:`Reclassification table` parameter.
  A table like dialog will pop up where you can choose the minimum, maximum and
  new values for each class.
* Click on the :guilabel:`Add row` button and add 5 rows. Fill each row as the
  following picture and click :guilabel:`OK`:

  .. image:: img/reclassify_table.png
     :align: center

* The method used by the algorithm to treat the threshold values of each class
  is defined by the :guilabel:`Range boundaries` parameter;
* Save the layer as ``reclassified`` in the ``exercise_data/raster_analysis/``
  folder;

  .. image:: img/reclassify_setup.png
     :align: center

* Click on :guilabel:`Run`.

If you compare the native :guilabel:`aspect` layer with the
:guilabel:`reclassified` one, there are not big differences. But giving a look
at the legend you can see that the values go from 1 to 4.

> Let's give this layer a better style. Open the :guilabel:`Layer Styling` panel

Reviewer 3: Use list for these steps, ie 

> Let's give this layer a better style.
>
> #. Open the :guilabel:`Layer Styling` panel ...

and choose :guilabel:`Paletted/Unique values` instead of :guilabel:`Singleband gray`.
Click on the :guilabel:`Classify` button to automatically fetch the values and
assign them random colors:

.. image:: img/unique_style.png
   :align: center

The output should look like this (you can have different colors given that they
have been randomly generated):

.. image:: img/reclassify_result.png
   :align: center

With this reclassification and the paletted style applied to the layer you can
immediately see the aspect areas. Cool isn't it?!

|basic| |FA| Querying the raster
-------------------------------------------------------------------------------

Unlike vectors, raster layers don't have an attribute table: each pixel contains
one or more numerical values, depending if the raster is singleband or multiband.

All the raster layers we used in this exercise are made by just a single band:
depending on the layer, pixel numbers will represent elevation, aspect or slope
values.

> How can we query the raster layer to now the value of a single pixel? We can use

Reviewer 3: ```suggestion
How can we query the raster layer to know the value of a single pixel? We can use
```

the |identify| button to extract this information.

Select the tool from the upper toolbar and click on a random location of the
:guilabel:`srtm_41_19` layer. The :guilabel:`Identify Results` will appear
with the value of the band at the clicked location:

.. image:: img/identify_raster.png
   :align: center

You can change the view of the :guilabel:`Identify Results` panel from the current

> ``tree`` view to a ``table`` view by selecting ``Table`` in the :guilabel:`View`

Reviewer 3: A lot of "view" in this sentence :smiley:
Should also use :guilabel: here

menu at the bottom of the panel:

.. image:: img/identify_raster_table.png
   :align: center

Clicking each pixel to get the value of the raster could become annoying after
a while. We can use the :guilabel:`Value Tool` plugin to solve this problem.

#. Go to :menuselection:`Plugins --> Manage/Install Plugins...`
#. In the :guilabel:`All` tab, type ``Value Tool`` in the search box;
#. Select the Value Tool plugin, press :guilabel:`Install Plugin` and then
   :guilabel:`Close` the dialog;

   .. image:: img/value_tool.png
      :align: center

The new :guilabel:`Value Tool` panel will appear.

.. tip:: If you close the panel you can reopen it by enabling it in the
    :menuselection:`View --> Panels --> Value Tool` or by clicking on the new
    icon of the toolbar.

To use the plugin just check the :guilabel:`Enable` checkbox and be sure that
the :guilabel:`srtm_41_19` layer is active (checked) in the :guilabel:`Layers`
panel.

> Move the cursor on the map to immediately now the value of the pixel.

Reviewer 3: ```suggestion
Move the cursor on the map to immediately know the value of the pixel.
```

> .. image:: img/value_too_query.png

Reviewer 3: ```suggestion
.. image:: img/value_tool_query.png
```
also rename the file
